File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int int_length(long int);
int get_first_digits(long int, int);

int main(void)
{
    long int card_number = get_long("What is your card number? ");

    long int n;
    int mult_digit2;
    int sum_mult_digit2 = 0;
    int sum_digit1 = 0;
    for (n = 10; n <= 1000000000000000; n *= 100)
    {
        // multiply every 2nd digit (counting from the 2nd from the end) by 2
        mult_digit2 = (floor((card_number % (n * 10)) / n)) * 2;
        // add those mult_digits2 together (not the numbers, but their digits)
        sum_mult_digit2 += (mult_digit2 % 10) + (floor(mult_digit2 / 10));

    }

    // sum of every 2nd digit (counting from the 1st!!! from the end)
    for (n = 1; n <= 100000000000000; n *= 100)
    {
        sum_digit1 = sum_digit1 + (floor(card_number % (n * 10)) / n);
    }

    //card number length.
    int length = int_length(card_number);
    //first two numbers check
    int first_two_digits = get_first_digits(card_number, 2);
    //first number check
    int first_digit = get_first_digits(card_number, 1);
    //check is the second digit after the multiplication and addition is 0
    bool is_card_hash_valid = (sum_mult_digit2 + sum_digit1) % 10 == 0;

    // define what type of a card is it
    if (!is_card_hash_valid)
    {
        printf("INVALID\n");
    }
    else if (length == 15 && (first_two_digits == 34 || first_two_digits == 37))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && first_two_digits >= 51 && first_two_digits <= 55)
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && first_digit == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

// a function that gets length of a long int variable
int int_length(long int x)
{
    int length = 0;
    while (x > 0)
    {
        x /= 10;
        length += 1;
    }
    return length;
}

// a funtion that 
int get_first_digits(long int x, int digits)
{
    int length = int_length(x);
    return floor(x / pow(10, length - digits));
}

File: sentimental-credit/credit.py
# TODO
from cs50 import get_string

def main():
# promt user for a card number
    card_number = get_string("Number: ")

# check if it is a valid number
    if not validation_check(card_number):
        print("INVALIDff")
        return 1

# check whether it is an American Express, MasterCard, or Visa card number
    if amex(card_number):
        print("AMEX")
        return 0
    if mastercard(card_number):
        print("MASTERCARD")
        return 0

    if visa(card_number):
        print("VISA")
        return 0
    else:
        print("INVALID")
    return 0




def validation_check(x):
# we take every other digit, starting with the number’s second-to-last digit:
# multiply each of the digits by 2
# add those products’ digits (i.e., not the products themselves) together
# add that sum to the sum of the remaining digits (that weren’t multiplied by 2)
    sum = 0
    for i in x[-2::-2]:
        multiplied_digit = int(i) * 2
        sum += multiplied_digit % 10
        sum += multiplied_digit // 10
    for i in x[-1::-2]:
        sum += int(i)
# check if the last digit is 0
    if sum % 10 == 0:
        return True
    else:
        return False


def amex(a):
# check length of the card number
    length = len(a)
    if length == 15:
        return True
# check the first two digits
    if a[0] != 3:
        return False
    if a[1] != 4 or a[1] != 7:
        return False
    else:
        return True

def mastercard(a):
# check length of the card number
    length = len(a)
    if length != 16:
        return False

# check the first two digits
    if a[0] != 5:
        return False
    if a[1] != 1 or a[1] != 5:
        return False
    else:
        return True

def visa(a):
# check length of the card number
    length = len(a)
    if length != 13 or length != 16:
        return False

# check the first digit
    if a[0] != 4:
        return False
    else:
        return True
main()


